### PR TITLE
Add DID DHT method

### PR DESCRIPTION
**Pull request by decentralgabe**
https://github.com/TBD54566975/did-dht-method

### CODE REVIEW

**Comment by decentralgabe** [comment]
seeing ```error TS2209: The project root is ambiguous, but is required to resolve export map entry '.' in file '/home/runner/work/web5-js/web5-js/packages/dids/package.json'. Supply the `rootDir` compiler option to disambiguate.```

🤔

File: package.json
Comment (decentralgabe) [inline package.json:28]: `npx depcheck` complained about these, but can remove them

**Comment by decentralgabe** [comment]
updated to use dns packets as per https://hackmd.io/-fBsfngHRpuJilwf5GY3Qg?both#DIDs-as-a-DNS-Packet

cc: @csuwildcat @Nuhvi

File: packages/dids/src/dht.ts
Comment (mistermoe) [inline packages/dids/src/dht.ts:4]:
```suggestion
import type {DidDocument} from './types.js';
```

**Comment by mistermoe** [comment]
@decentralgabe convention for imports in this repo is:

1. external dep types
2. external deps
3. internal dep types
4. internal deps

[example](https://github.com/TBD54566975/web5-js/blob/main/packages/dids/src/did-ion.ts#L1-L12)

anything i'm missing w.r.t. to imports @frankhinek ?

File: packages/dids/.mocharc.json
Comment (mistermoe) [inline packages/dids/.mocharc.json:8]: any specific reason for using `ts-node`? we've gotten by without having to use it so far.

File: packages/dids/src/dht.ts
Comment (mistermoe) [inline packages/dids/src/dht.ts:1]: I think `PublicKeyJwk` and `Web5Crypto` are types which we can import explicitly as types.

File: packages/dids/.mocharc.json
Comment (decentralgabe) [reply packages/dids/.mocharc.json:8]: it lets me run mocha tests in webstorm really simply

happy to remove it and just re-add locally / remove before i push if easier

File: packages/dids/src/did-dht.ts
Comment (mistermoe) [inline packages/dids/src/did-dht.ts:2]: `crypto` isn't available in the browser unless we polyfill it. AFAICT you're only using `crypto` to generate digests which we're using [`@noble/hashes`](https://www.npmjs.com/package/@noble/hashes) for. [examples here](https://github.com/search?q=repo%3ATBD54566975%2Fweb5-js+%40noble%2Fhashes%2Fsha&type=code)

@frankhinek, i forget, did we end up deciding it'd be worth adding common hashing functions to `common`?

File: packages/dids/src/did-dht.ts
Comment (mistermoe) [inline packages/dids/src/did-dht.ts:163]: nothing immediate just wanted to call out that the same code exists [here](https://github.com/TBD54566975/web5-js/blob/main/packages/dids/src/did-key.ts#L644-L666) for `did:key` and [here](https://github.com/TBD54566975/web5-js/blob/f3c7184bab45972821c89d6275eda8a6c295884f/packages/dids/src/did-ion.ts#L260-L282) for `did:ion`

@frankhinek thoughts on eventually adding a `Web5SubleCrypto` api surface that matches [W3C WebCrypto's SubtleCrypto](https://www.w3.org/TR/WebCryptoAPI/#subtlecrypto-interface)?

File: packages/dids/.mocharc.json
Comment (mistermoe) [reply packages/dids/.mocharc.json:8]: ah gotcha. did you have to configure something specific to get things to work with webstorm? i'm thinking we can get webstorm to continue working really simply without introducing `ts-node`. 

p.s. we should move the `ts-node` dependency to `devDependencies` in `package.json`

File: packages/dids/src/did-dht.ts
Comment (decentralgabe) [reply packages/dids/src/did-dht.ts:2]: worked for me

File: packages/dids/src/dht.ts
Comment (decentralgabe) [reply packages/dids/src/dht.ts:1]: cool didn't know about type imports, what's it do?

File: packages/dids/.mocharc.json
Comment (decentralgabe) [reply packages/dids/.mocharc.json:8]: yeah had to manually add a run config, happy to work with you to find another solution

**Comment by mistermoe** [comment]
kk @decentralgabe it looks like `bittorrent-dht` and `bittorrent-dht-sodium` are not browser compatible. 

`bittorrent-dht` relies on:
* `crypto`
* `dgram`
* `fs`
* `net`
* `dns`
* `path`
* `os`

if `bittorrent-dht` is a client used to send UDP requests, we won't be able to polyfill our way through getting it to work anyway because browsers can't make UDP requests. we'd need to go through an http gateway

`bittorent-dht-sodium` relies on:
* `node-gyp`

 it's using `node-gyp` to wrap `libsodium` and surface it to nodejs. we have the necessary primitives needed to hand roll this  given all of the work @frankhinek has done on the Ed25519 front. 

from what i can tell (could be wrong)`bittorrent-dht-sodium` isn't [doing much](https://github.com/emilbayes/bittorrent-dht-sodium/blob/master/index.js) aside from generating Ed25519 keypairs and providing sign/verify functions (all of which we have). what may be tricky is supporting this:

```js
const dht = new DHT({ verify: eddsa.verify })
```

given that `eddsa.verify` is coming out of `bittorrent-dht-sodium`. will have to go look at the function signature of `eddsa.verify` to see if we can provider our own `verify` function. really hoping its `Uint8Array` in / `Uint8Array` out

cc: @frankhinek @csuwildcat

**Comment by mistermoe** [comment]
p.s. CI failing now bc of linting issues in `dht.js`. should be able to configure webstorm to autofix on save which will take care of the spacing issues

**Comment by csuwildcat** [comment]
@decentralgabe we definitely need to get this browser and React Native compatible before pulling it in.

File: package.json
Comment (decentralgabe) [reply package.json:1]: thanks

**Comment by decentralgabe** [comment]
without `bittorrent-dht` we lose the ability to publish/read all together. if this is necessary we'll need to rely on our own gateways which surface a traditional HTTP API over using the DHT directly, which is not ideal.

**Comment by decentralgabe** [comment]
updated to use the `pkarr` lib and remove reliance on `bittorrent-dht` direclty

File: packages/dids/build/buffer-polyfill.cjs
Comment (decentralgabe) [inline packages/dids/build/buffer-polyfill.cjs:1]: @kirahsapong @mistermoe did this for the pollyfill and it works great

**Comment by decentralgabe** [comment]
ready to go pending final review/approval by @mistermoe @frankhinek

**Comment by mistermoe** [comment]
@frankhinek any reservations on pulling this in?

**Comment by frankhinek** [comment]
Just started review.  Also asked @shamilovtim to take a look to see what might need to be done to ensure that the `@web5/dids` package is still usable under React Native after this is merged.

**Comment by shamilovtim** [comment]
I'll be taking a peak at this ASAP 👍

related: https://github.com/TBD54566975/web5-js/pull/240

File: packages/dids/package.json
Comment (frankhinek) [inline packages/dids/package.json:80]: @decentralgabe any particular reason to list `bittorrent-dht` as an explicit dependency?  It appears to be a transitive dependency of the `pkarr` package.  Could lead to version conflicts to have it specified in `@web5/dids`.

File: packages/dids/package.json
Comment (frankhinek) [inline packages/dids/package.json:81]: @decentralgabe any particular reason to list `bittorrent-dht-sodium` as an explicit dependency?  It doesn't appear to be used anywhere in the PR nor is it a dependency of any other packages as best I can tell.

File: packages/dids/package.json
Comment (frankhinek) [inline packages/dids/package.json:109]: @decentralgabe why were all of these dev dependencies removed?

File: packages/dids/src/dht.ts
Comment (frankhinek) [inline packages/dids/src/dht.ts:25]: @decentralgabe should the pkarr relay used to publish be a configurable parameter?  Fine with addressing this in a future PR.

File: packages/dids/src/dht.ts
Comment (frankhinek) [inline packages/dids/src/dht.ts:175]: @decentralgabe should the pkarr relay used to retrieve the DID doc be a configurable parameter?  Fine with addressing this in a future PR.

**Review by frankhinek — CHANGES_REQUESTED** [review]

File: packages/dids/package.json
Comment (decentralgabe) [reply packages/dids/package.json:109]: I ran `npx depcheck` and it told me to remove them - can revert if you want

File: packages/dids/package.json
Comment (decentralgabe) [reply packages/dids/package.json:80]: remnant, removing

File: packages/dids/package.json
Comment (decentralgabe) [reply packages/dids/package.json:81]: remnant, removing

File: packages/dids/package.json
Comment (decentralgabe) [reply packages/dids/package.json:79]: remnant, removing

File: packages/dids/src/dht.ts
Comment (decentralgabe) [reply packages/dids/src/dht.ts:25]: I believe it's something the library should take a stance on, but may be able to override

File: packages/dids/src/dht.ts
Comment (decentralgabe) [reply packages/dids/src/dht.ts:175]: same comment as above

Comment (decentralgabe) [reply packages/dids/src/dht.ts]: updated to support a configurable relay with a default set by the lib

File: packages/dids/src/dht.ts
Comment (decentralgabe) [reply packages/dids/src/dht.ts]: updated to support a configurable relay with a default set by the lib

**Review by frankhinek — APPROVED** [review]

**Pull request merged**
